### PR TITLE
Design of creating user defined multiple database instances

### PR DESCRIPTION
**Pull request by dzhangalibaba**
* Detail explanation of the design for creating user defined multiple database instances 
*  Pull request for implementation is https://github.com/Azure/sonic-buildimage/pull/2182

Signed-off-by: Dong Zhang d.zhang@alibaba-inc.com

### CODE REVIEW

**Comment by jipanyang** [comment]
you mentioned that to use redis cluster solution,  client has to use TCP + PORT instead of unix socket,  do you have some rough single client read/write performance benchmark for the two different connections against single redis instance?

**Comment by dzhangalibaba** [comment]
> you mentioned that to use redis cluster solution, client has to use TCP + PORT instead of unix socket, do you have some rough single client read/write performance benchmark for the two different connections against single redis instance?

update  the benchmark result in DOC

File: doc/database/multi_database_instances.md [hunk 1]
```diff
@@ -0,0 +1,506 @@
...
+   - [x] if no folder /host/old_config/, copy some default xmls and etc. as usual
+3. **database service**
+   - [x] **make database service depends on rc.local service since database needs to access old_config/config_db.json to get DATABASE configuration earlier**
+   - [x] **database.sh start**
+     - [x] **access and copy /etc/sonic/old_config/database_config.json into /etc/sonic/**
+     - [x] **If there is no old_config folder or database_config.json file, we take it as no extra DATABASE requirement  and create a empty "{}" database_config.json.**
+     - [x] **generate corresponding runtime ping/PONG check script as well to check if database instances are running later**
+   - [x] **docker ENTRYPOINT : docker_init.sh**
+     - [x] **at this point, we know the DATABASE configuration in database_config.json**
+     - [x] **generate supervisord.conf and all redis.conf before database docker start**
```
[the remark this answers is not in the document]
Comment (dzhangalibaba) [reply doc/database/multi_database_instances.md]: updated

File: doc/database/multi_database_instances.md [hunk 2]
```diff
@@ -0,0 +1,506 @@
...
+  ```json
+  {
+      "DATABASE": {
+          "redis_ins_001":{
+              "port": 6380,
+              "databases" : ["LOGLEVEL_DB","SYSMON_DB","COUNTERS_DB"]
+          },
+          "redis_ins_002":{
+              "port": 6381,
+              "databases" : ["APPL_DB","ASIC_DB", "CONFIG_DB"]
```
[the remark this answers is not in the document]
Comment (dzhangalibaba) [reply doc/database/multi_database_instances.md]: updated

**Comment by qiluo-msft** [comment]
For backward compatibility, please keep the option to specify unix socket for each redis instance. We prefer it for performance issue.

**Comment by dzhangalibaba** [comment]
> For backward compatibility, please keep the option to specify unix socket for each redis instance. We prefer it for performance issue.

we have uinx socket, by default each inistance has a unix socket at "/var/run/instance_name/instance_name.sock",
for example，
redis ： "/var/run/redis/redis.sock"
redis_ins_2: "/var/run/redis_ins_2/redis_ins_2.sock"
...


you mean the user define the unix socket path and name in database_config.json file ?

**Comment by qiluo-msft** [comment]
Thanks! Reading through the document, I am not aware this design detail.

---
In reply to: [481505546](https://github.com/Azure/SONiC/pull/271#issuecomment-481505546) [](ancestors = 481505546)

**Pull request merged**